website/addons/googledrive/model.py

```diff
@@ -2,235 +2,189 @@
-        if self.oauth_settings:
-            return self.oauth_settings.refresh_token
-        return None
+    drive_folder_id = fields.StringField(default=None)
+    drive_folder_name = fields.StringField(default=None)
+    folder_path = fields.StringField()
+    serializer = GoogleDriveSerializer
 
-    @refresh_token.setter
-    def refresh_token(self, val):
-        self.oauth_settings.refresh_token = val
+    _api = None
 
     @property
-    def expires_at(self):
-        if self.oauth_settings:
-            return self.oauth_settings.expires_at
-        return None
-
-    @expires_at.setter
-    def expires_at(self, val):
-        self.oauth_settings.expires_at = val
+    def api(self):
+        """Authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = GoogleDriveProvider(self.external_account)
+            #self._api.account = self.external_account
+        return self._api
 
     @property
-    def has_auth(self):
-        if self.oauth_settings:
-            return self.oauth_settings.access_token is not None
-        return False
-
-    def fetch_access_token(self, force_refresh=False):
-        if self.oauth_settings:
-            return self.oauth_settings.fetch_access_token(force_refresh)
-        return None
-
-    def clear(self):
-        if self.oauth_settings:
-            self.oauth_settings.revoke_access_token()
-            self.oauth_settings = None
-            self.save()
-
-        for node_settings in self.googledrivenodesettings__authorized:
-            node_settings.deauthorize(Auth(self.owner))
-            node_settings.save()
-
-    def save(self, *args, **kwargs):
-        if self.oauth_settings:
-            self.oauth_settings.save()
-        return super(GoogleDriveUserSettings, self).save(*args, **kwargs)
-
-    def delete(self, save=True):
-        self.clear()
-        super(GoogleDriveUserSettings, self).delete(save)
-
-    def __repr__(self):
-        return u'<GoogleDriveUserSettings(user={self.owner.username!r})>'.format(self=self)
-
-
-class GoogleDriveNodeSettings(StorageAddonBase, AddonNodeSettingsBase):
-
-    folder_id = fields.StringField(default=None)
-    folder_path = fields.StringField(default=None)
-
-    user_settings = fields.ForeignField(
-        'googledriveusersettings', backref='authorized'
-    )
+    def complete(self):
+        return bool(self.has_auth and self.user_settings.verify_oauth_access(
+            node=self.owner,
+            external_account=self.external_account,
+            metadata={'folder': self.drive_folder_id}
+        ))
 
-    @property
     def folder_name(self):
-        if not self.folder_id:
+        if not self.drive_folder_id:
             return None
 
         if self.folder_path != '/':
-            return unquote(os.path.split(self.folder_path)[1].encode('utf-8')).decode('utf-8')
-
-        return '/ (Full Google Drive)'
+            # `urllib` does not properly handle unicode.
```

Comment:
Thanks. Did we establish why we were using urllib on the folder path in the first place? 

(I suspect this is something to do with the folder data google sends back, but don't want to take that assumption for granted)